src/manager/checkpoint/manager/bottomup.rs

```diff
@@ -1,49 +1,288 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+use crate::config::Subnet;
+use crate::lotus::client::DefaultLotusJsonRPCClient;
+use crate::lotus::message::mpool::MpoolPushMessage;
+use crate::lotus::LotusClient;
 use crate::manager::checkpoint::CheckpointManager;
+use anyhow::anyhow;
 use async_trait::async_trait;
+use cid::Cid;
+use fil_actors_runtime::cbor;
 use fvm_shared::address::Address;
 use fvm_shared::clock::ChainEpoch;
+use fvm_shared::MethodNum;
 use ipc_gateway::BottomUpCheckpoint;
 use ipc_sdk::subnet_id::SubnetID;
+use primitives::TCid;
+use std::str::FromStr;
 
-pub struct BottomUpCheckpointManager;
+pub struct BottomUpCheckpointManager {
+    parent: SubnetID,
+    parent_client: DefaultLotusJsonRPCClient,
+    child_subnet: Subnet,
+    child_client: DefaultLotusJsonRPCClient,
+
+    checkpoint_period: ChainEpoch,
+}
+
+impl BottomUpCheckpointManager {
+    pub fn new_with_period(
+        parent_subnet: SubnetID,
+        parent_client: DefaultLotusJsonRPCClient,
+        child_subnet: Subnet,
+        child_client: DefaultLotusJsonRPCClient,
+        checkpoint_period: ChainEpoch,
+    ) -> Self {
+        Self {
+            parent: parent_subnet,
+            parent_client,
+            child_subnet,
+            child_client,
+            checkpoint_period,
+        }
+    }
+
+    pub async fn new(
+        parent_client: DefaultLotusJsonRPCClient,
+        parent: SubnetID,
+        child_client: DefaultLotusJsonRPCClient,
+        child_subnet: Subnet,
+    ) -> anyhow::Result<Self> {
+        let checkpoint_period = obtain_checkpoint_period(&child_subnet.id, &parent_client).await?;
+        Ok(Self::new_with_period(
+            parent,
+            parent_client,
+            child_subnet,
+            child_client,
+            checkpoint_period,
+        ))
+    }
+
+    async fn last_executed_epoch(&self) -> anyhow::Result<ChainEpoch> {
+        let parent_head = self.parent_client.chain_head().await?;
+
+        // A key assumption we make now is that each block has exactly one tip set. We panic
+        // if this is not the case as it violates our assumption.
+        // TODO: update this logic once the assumption changes (i.e., mainnet)
+        assert_eq!(parent_head.cids.len(), 1);
+
+        let cid_map = parent_head.cids.first().unwrap().clone();
+        let parent_tip_set = Cid::try_from(cid_map)?;
+        // get subnet actor state and last checkpoint executed
+        let subnet_actor_state = self
+            .parent_client
+            .ipc_read_subnet_actor_state(&self.child_subnet.id, parent_tip_set)
+            .await?;
+
+        Ok(subnet_actor_state
+            .bottom_up_checkpoint_voting
+            .last_voting_executed)
+    }
+}
 
 #[async_trait]
 impl CheckpointManager for BottomUpCheckpointManager {
-    type Checkpoint = BottomUpCheckpoint;
+    async fn obtain_validators(&self) -> anyhow::Result<Vec<Address>> {
```

Comment:
Same here, we should maybe add the tipset as an argument and that way we can get the tipset once and get the data that we need from it.

Reply:
see comment above